okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ConcurrencyLimiters.java

```diff
@@ -0,0 +1,147 @@
+    private static final Logger log = LoggerFactory.getLogger(ConcurrencyLimiters.class);
+    private static final String FALLBACK = "";
+    private final ConcurrentMap<String, AsyncLimiter> limiters = new ConcurrentHashMap<>();
+
+    private final Map<Limiter.Listener, Runnable> activeListeners = CacheBuilder.newBuilder()
+            .weakKeys()
+            .<Limiter.Listener, Runnable>removalListener(notification -> {
+                if (notification.getCause().equals(RemovalCause.COLLECTED)) {
+                    log.warn("Concurrency limiter was leaked."
+                            + " This implies a remoting bug or classpath issue, and may cause degraded performance");
+                    notification.getValue().run();
+                }
+            })
+            .build()
+            .asMap();
+
+    @VisibleForTesting
+    AsyncLimiter limiter(String name) {
+        return limiters.computeIfAbsent(name, key ->
+                new AsyncLimiter(activeListeners, DefaultLimiter.newBuilder()
+                        .limit(TracingLimitDecorator.wrap(VegasLimit.newBuilder().initialLimit(1).build()))
+                        .build(new SimpleStrategy<>())));
+    }
+
+    public AsyncLimiter limiter(Request request) {
+        final String limiterKey;
+        String pathTemplate = request.header(OkhttpTraceInterceptor.PATH_TEMPLATE_HEADER);
+        if (pathTemplate == null) {
+            limiterKey = FALLBACK;
+        } else {
+            limiterKey = request.method() + " " + pathTemplate;
+        }
+        return limiter(limiterKey);
+    }
+
+    static final class AsyncLimiter {
+        private final Map<Limiter.Listener, Runnable> activeListeners;
+        private final Queue<SettableFuture<Limiter.Listener>> waitingRequests = new LinkedBlockingQueue<>();
+        private final Limiter<Void> limiter;
+
+        public AsyncLimiter(
+                Map<Limiter.Listener, Runnable> activeListeners,
+                Limiter<Void> limiter) {
+            this.activeListeners = activeListeners;
+            this.limiter = limiter;
+        }
+
+        public ListenableFuture<Limiter.Listener> acquire() {
+            Optional<Limiter.Listener> maybeListener = limiter.acquire(null);
+            if (maybeListener.isPresent()) {
+                return Futures.immediateFuture(wrap(activeListeners, maybeListener.get()));
+            }
+            SettableFuture<Limiter.Listener> future = SettableFuture.create();
+            waitingRequests.add(future);
+            return future;
+        }
+
+        private void processQueue() {
+            while (!waitingRequests.isEmpty()) {
+                Optional<Limiter.Listener> maybeAcquired = limiter.acquire(null);
+                if (!maybeAcquired.isPresent()) {
+                    return;
+                }
+                Limiter.Listener acquired = maybeAcquired.get();
+                SettableFuture<Limiter.Listener> head = waitingRequests.poll();
```

Comment:
would prefer to make the multi-threadedness here a little easier to understand by calling `poll` only
```
SettableFuture<> head;
// Note that different threads may be executing processQueue; this is safe because ...
while( (head = waitingRequests.poll()) != null ) {
  ..
}
```

Reply:
ah but if you do that, you're not guaranteed to have any permits to give them

Reply:
i see. then add a comment, please

Reply:
done